scheduler-server/scheduling-dsl-compiler/src/libs/windows-edsl-fluent-api.ts

```diff
@@ -0,0 +1,107 @@
+
+export class WindowSet {
+  public readonly __astnode: AST.WindowsExpression;
+  private constructor(windowSpecifer: AST.WindowsExpression) {
+    this.__astnode = windowSpecifer;
+  }
+  private static new(windowSpecifier: AST.WindowsExpression) {
+    return new WindowSet(windowSpecifier);
+  }
+
+  public and(...others: WindowSet[]): WindowSet {
+    return WindowSet.new({
+      kind: AST.NodeKind.WindowsExpressionAnd,
+      windowsExpressions: [
+        this.__astnode,
+        ...others.map(other => other.__astnode),
+      ],
+    });
+  }
+
+  public or(...others: WindowSet[]): WindowSet {
+    return WindowSet.new({
+      kind: AST.NodeKind.WindowsExpressionOr,
+      windowsExpressions: [
+        this.__astnode,
+        ...others.map(other => other.__astnode),
+      ],
+    });
+  }
+
+  public static gt(resource: Resource, value: Double): WindowSet {
+    return WindowSet.new({
+      kind: AST.NodeKind.WindowsExpressionGreaterThan,
+      left: resource,
+      right: value,
+    });
+  }
+
+  public static lt(resource: Resource, value: Double): WindowSet {
+    return WindowSet.new({
+      kind: AST.NodeKind.WindowsExpressionLessThan,
+      left: resource,
+      right: value,
+    });
+  }
+
+  public static eq(resource: Resource, value: Double): WindowSet {
+    return WindowSet.new({
+      kind: AST.NodeKind.WindowsExpressionEqualLinear,
+      left: resource,
+      right: value,
+    });
+  }
+
+  public static neq(resource: Resource, value: Double): WindowSet {
+    return WindowSet.new({
+      kind: AST.NodeKind.WindowsExpressionNotEqualLinear,
+      left: resource,
+      right: value,
+    });
+  }
+
+  public static between(resource: Resource, lowerBound: Double, upperBound: Double): WindowSet {
+    return WindowSet.new({
+      kind: AST.NodeKind.WindowsExpressionBetween,
+      resource,
+      lowerBound,
+      upperBound
+    });
+  }
+
+  public static transition(resource: Resource, from: any, to: any): WindowSet {
+    return WindowSet.new({
+      kind: AST.NodeKind.WindowsExpressionTransition,
+      resource,
+      from,
+      to
+    });
+  }
+
+  public static during(activityType: ActivityType): WindowSet {
+    return WindowSet.new({
+      kind: AST.NodeKind.ActivityExpression,
+      type: activityType
+    })
+  }
+}
+
+declare global {
+  class WindowSet {
+    public readonly __astnode: AST.WindowsExpression
+    public and(...others: WindowSet[]): WindowSet
+    public or(...others: WindowSet[]): WindowSet
+    public static gt(resource: Resource, value: Double): WindowSet
+    public static lt(resource: Resource, value: Double): WindowSet
+    public static eq(resource: Resource, value: Double): WindowSet
+    public static neq(resource: Resource, value: Double): WindowSet
+    public static between(resource: Resource, lowerBound: Double, upperBound: Double): WindowSet
+    public static transition(resource: Resource, from: any, to: any): WindowSet
```

Comment:
the any types here are a bit concerning. Is there anyway we can make this a type that we replace with generated code at runtime? Like we do with ActivityTypes and Resources

Reply:
Latest commit attempts to address this problem using function overloading.

I don't want to delve too deep into it in this PR, since @JoelCourtney's work parallels this. If we decide this is a good way to do things, hopefully we can reuse some of these ideas on the Constraints side.

Reply:
After a discussion with @dyst5422 and @adrienmaillard , I've updated this to use [Conditional Types](https://www.typescriptlang.org/docs/handbook/2/conditional-types.html) instead of function overloading. Let me know what you think!

I'll squash the last two commits before merging this PR, since one of them uses function overloading, and the other uses conditional types.

I'll echo the sentiment in my previous comment that this code will need to move out of the scheduler and into the Constraints eDSL during the next sprint.